### DISCUSSION TITLE
Graph sampling conditional on node/edge attr

### DISCUSSION THREAD

**Discussion by ssharpe42** [pyg-team/pytorch_geometric#3292] [Q&A]
Are there any existing implementations or suggestions for node samplers where you can sample time directed subgraphs. 

For example sample neighbors based on edge/node values so that all k hop neighbors `v` of node `u` have `t_v < t_u`

**Comment by rusty1s** [comment]
Such functionality is sadly not supported yet, but it would be quite important to add. I'm not entirely sure how the concrete interface should look like though. Shall `t` be given as a static vector during initialization of the sampler, or might it be important to apply such masking of nodes dynamically?

**Reply by ssharpe42** [reply]
Im envisioning this as a way to allow for training static GNN models with a single snapshot of a graph, but only allowing relevant information to be available for each node's neighborhood. Since `t` is fixed in a static graph (nodes/edges appear at time t_i) what do you mean by masking nodes dynamically?

Do you mean for example masking 2nd hop neighbors based on the directly sampled neighbors?  I think there are two possibilities here: 
1. A strict time directed graph sampling where each sampled neighborhood forms a directed graph that points towards the target node
2. A graph where all edges and nodes occur before the target node. This differs from 1 where a 1st hop neighbor occurs at t=2, 2nd hop neighbor occurs at t=4 and target node occurs at t=6. Not a time directed graph, but all multihop neighbors of the target node are present at t=6.

**Reply by rusty1s** [reply]
I see. I personally think that the second option makes more sense. In that case, sampling would have an interface similar to:
```python
TemporalNeighborLoader(data, ..., temporal_attr='node_timestamp')
```
where `temporal_attr` points to the attribute in `data` that holds temporal information (either node-level or edge-level vector). I think the `TemporalNeighborLoader` would need to be different than our default `NeighborLoader`, as we need distinct subgraphs for each target node to prevent data leakage across target nodes.
